pkg/dataplane/client/xds.go

```diff
@@ -0,0 +1,76 @@
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/clusterlink-net/clusterlink/pkg/dataplane/server"
+)
+
+// XDSClient implements the client which fetches clusters and listeners
+type XDSClient struct {
+	dataplane          *server.Dataplane
+	controlplaneTarget string
+	tlsConfig          *tls.Config
+	fetchers           []*fetcher
+	errors             []error
+	logger             *logrus.Entry
+}
+
+func (x *XDSClient) runFetcher(_ *fetcher, resourceType string) error {
+	for {
+		time.Sleep(5 * time.Second)
+		conn, err := grpc.Dial(x.controlplaneTarget, grpc.WithTransportCredentials(credentials.NewTLS(x.tlsConfig)))
+		if err != nil {
+			x.logger.Errorf("Failed to dial controlplane xDS server: %v.", err)
+			continue
+		}
+		x.logger.Infof("Successfully connected to the controlplane xDS server.")
+
+		f, err := newFetcher(context.Background(), conn, resourceType, x.dataplane)
+		if err != nil {
+			x.logger.Errorf("Failed to initialize fetcher: %v.", err)
+			continue
+		}
+		x.logger.Infof("Successfully initialized client for %s type.", resourceType)
+		err = f.Run()
+		x.logger.Infof("Fetcher '%s' stopped: %v.", resourceType, err)
+	}
+}
+
+// Run starts the running xDS client which fetches clusters and listeners from the controlplane.
+func (x *XDSClient) Run() error {
+	var wg sync.WaitGroup
+	wg.Add(len(x.fetchers))
+	go func() {
```

Comment:
suggestion:
consider using a for loop on elements of an array of `[resource.ClusterType, resource.ListenerType]`
and then you can fold the two go functions into a single code snippet.
If we add (e.g.,) `resource.SecretType` in the future, it becomes an easy addition with no code duplication